### PR TITLE
Hotfix/ehlpxt 315 tls vodafone connections leak

### PR DESCRIPTION
**Pull request by AndreyDodonov-EH**
**TL;DR**
Cherry-pick part of https://github.com/zephyrproject-rtos/zephyr/pull/57509, acknowledging data from the host which are received after we sent FIN.

**Problem summary**
It was observed that with some connections (e.g. via Vodafone SIM with Vodafone APN) device stops sending data after a couple of minutes.
Symptom is - can not open socket.

**Investigation**
Further debugging showed that all connections, i.e. sockets were in use, since they had not been closed.

Then wireshark traces were collected.
For that Vodafone SIM (not same as by client, different one) was inserted in Teltonika router. Router then was connected to a MAC, which created an own Wi-Fi network, to which device was connected.

Logs showed that connection was never closed, and stayed in TCP_FIN_WAIT_2 state - host was acknowledging our FIN, but kept sending data. So we never closed the connection, since we were waiting for FIN from the host.
So connections were leaked and eventually we couldn't open new ones.

**RFC Background**
Normally, in TCP client upon closing, i.e. issuing FIN, should be able to process data from the host.
It is so-called [Half-Closed connection in rfc9293](https://www.rfc-editor.org/rfc/rfc9293#name-half-closed-connections).
However, Zephyr does not support processing data after closing the connection.

According to the RFC, if client does not support Half-Closed connection, it should be issuing RST.
It is not done in Zephyr.

Neither was it ACK-ing new data, faking (mimicking) normal Half-Closed connection.

**Work done in Zephyr**
In the new version of Zephyr, https://github.com/zephyrproject-rtos/zephyr/releases/tag/v3.4.0, Zephyr started to fake normal Half-Closed connection. It is faking because data were not forward to the application, but at least they were ACK-ed, which lead to closing the connection. The change was introduced in refactoring FIN: https://github.com/zephyrproject-rtos/zephyr/pull/57509
So it works, device keeps sending data, which is great for us and for the clients.

However, RST still is not sent.
So I opened this PR (there you can also find Wireshark traces I collected) https://github.com/zephyrproject-rtos/zephyr/issues/61181

It lead to this PR (at the moment of writing it is a draft) https://github.com/zephyrproject-rtos/zephyr/pull/61353
It should fix closing behavior and RST in particular.

P.S. Both `3.4.0` and `main` (even more) in Zephyr contain many other important fixes for TCP, e.g.

- https://github.com/zephyrproject-rtos/zephyr/issues/53500
https://github.com/zephyrproject-rtos/zephyr/pull/53891 @kica-z  
They finally fixed `net_tcp: context->tcp == NULL` error message!!!

- https://github.com/zephyrproject-rtos/zephyr/pull/57995
https://github.com/zephyrproject-rtos/zephyr/pull/57202 (edited) 
two undefined behaviors

- https://github.com/zephyrproject-rtos/zephyr/issues/55641
https://github.com/zephyrproject-rtos/zephyr/pull/55775
Ignoring timeout in socket 

- https://github.com/zephyrproject-rtos/zephyr/issues/58828
https://github.com/zephyrproject-rtos/zephyr/pull/58905
Two possible crashes by dereferencing

- https://github.com/zephyrproject-rtos/zephyr/pull/60880
Fixing potential deadlock

And many others...

Which is SUPER important for a worldwide rollout, but it may also introduce new bugs, so we have to double down on testing!!!

### CODE REVIEW

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -2430,19 +2430,41 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 		} else if (th && FL(&fl, ==, ACK, th_seq(th) == conn->ack)) {
 			tcp_send_timer_cancel(conn);
 			next = TCP_FIN_WAIT_2;
+		} else if (th && (len > 0)) {
+			tcp_send_timer_cancel(conn);
```
Comment (kica-z) [inline subsys/net/ip/tcp.c:2434]: Why could we not use all the changes from TCP_FIN_WAIT_1 state? Is there some logic that's incompatible with our version?

**Review by kica-z — CHANGES_REQUESTED** [review]
LGTM. Only one question -> see comment.

File: subsys/net/ip/tcp.c (hunk 1 above)
Comment (AndreyDodonov-EH) [reply subsys/net/ip/tcp.c:2434]: @kica-z  
First, this  is from another PR `verdict = NET_OK;` (but this is only for statistics, if my understanding is correct, see 

Also, I cherry-picked only what was addressing our problem and what I understood, being afraid of side-effects.
Actually, before finding that PR, I partly implemented the fix myself, which was similar.

But in general you are right, we can cherry-pick the change completely.
I don't have a strong opinion here.

Comment (kica-z) [reply subsys/net/ip/tcp.c:2434]: i think especially with the new update in mind we should keep the merge as simple as possible. If we cherry-pick the entire change there will be no conflict later on.
But I think the current version also works fine.

@clamattia what do you think?

Comment (AndreyDodonov-EH) [reply subsys/net/ip/tcp.c:2434]: I tend to agree, but please note that this particular part will be again replaced in https://github.com/zephyrproject-rtos/zephyr/pull/61353 (but it comes after 3.4.0)

File: subsys/net/ip/tcp.c [hunk 2]
```diff
@@ -2430,19 +2430,40 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 		} else if (th && FL(&fl, ==, ACK, th_seq(th) == conn->ack)) {
 			tcp_send_timer_cancel(conn);
 			next = TCP_FIN_WAIT_2;
+		} else if (th && (len > 0)) {
+				tcp_out(conn, FIN | ACK);
```
Comment (ssharks) [inline subsys/net/ip/tcp.c]: This sends out a FIN ACK with the wrong sequence number, please take this line from my closed PR.

**Review by ssharks — COMMENTED** [review]
Nice to merge these changes in. A small but important comment

File: subsys/net/ip/tcp.c [hunk 3]
```diff
@@ -2430,19 +2430,41 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
...
+				/* Received FIN on FIN_WAIT_2, so cancel the timer */
+				k_work_cancel_delayable(&conn->fin_timer);
+
+				conn_ack(conn, + 1);
+				tcp_out(conn, ACK);
+				next = TCP_TIME_WAIT;
+			} else if (len > 0) {
+				/* Send out a duplicate ACK */
+				tcp_out(conn,  ACK);
+			}
```
Comment (ghost) [inline subsys/net/ip/tcp.c:2467]: Did I understand correctly, that you did not take in `verdict = NET_OK` from here https://github.com/zephyrproject-rtos/zephyr/pull/57509/files , because it needs changes we do not yet have in our current version?

File: subsys/net/ip/tcp.c (hunk 1 above)
Comment (ghost) [reply subsys/net/ip/tcp.c:2434]: For me either is fine.
I do not quite understand why you did not take the changes from `TCP_FIN_WAIT_1` from here: https://github.com/zephyrproject-rtos/zephyr/pull/57509/files ?

**Review by ghost — COMMENTED** [review]
Very nice catch @AndreyDodonov-EH :muscle: 

I think, if we want to take it for the next release, cherry-picking is the best approach. After that, I would prioritize updating zephyr and also keep an eye on the HEAD to observe the further development in the tcp stack.

File: subsys/net/ip/tcp.c (hunk 2 above)
Comment (ssharks) [reply subsys/net/ip/tcp.c]: It should be: `tcp_out_ext(conn, (FIN | ACK), NULL, conn->seq - 1);`

File: subsys/net/ip/tcp.c (hunk 3 above)
Comment (AndreyDodonov-EH) [reply subsys/net/ip/tcp.c:2467]: @clamattia Correction of setting the verdict for statistics was mostly done here https://github.com/zephyrproject-rtos/zephyr/pull/54711/files

Further PRs just try to follow the pattern

File: subsys/net/ip/tcp.c (hunk 2 above)
Comment (AndreyDodonov-EH) [reply subsys/net/ip/tcp.c]: @ssharks I totally forgot our fork was public LOL
Thanks a lot for dropping by :-) 

Ah yes, that's from your new PR.
This affects only our SEQ number, not the number we are ACK-ing, correct?

Comment (ssharks) [reply subsys/net/ip/tcp.c]: It indeed only affects the SEQ number, which is important as otherwise the receiving side gets out of sync

Comment (AndreyDodonov-EH) [reply subsys/net/ip/tcp.c]: So the receiving side (not zephyr) then  ACK-s wrong sequence number?
But practically it's not relevant if we anyway sent out FIN?

Tests show that it still works, at least with the hosts I had, which might not be representable, of course

File: subsys/net/ip/tcp.c (hunk 3 above)
Comment (ghost) [reply subsys/net/ip/tcp.c:2467]: Thanks for the explanation

**Comment by AndreyDodonov-EH** [comment]
@clamattia @kica-z 
So, should I merge it like it is now?

**Review by ghost — APPROVED** [review]
